### PR TITLE
Move invite code functionality from cookie to session

### PR DESCRIPTION
**Pull request by mdtro**
Moves the invite functionality from cookies to the session. This is to harden the security of the platform.

With the cookie approach, a client can manipulate the cookie value for `pending-invite` resulting in situations where an invite code can be reused.

### CODE REVIEW

File: src/sentry/api/endpoints/api_tokens.py [hunk 1]
```diff
@@ -62,4 +62,11 @@ def delete(self, request: Request):
 
         ApiToken.objects.filter(user=request.user, token=token, application__isnull=True).delete()
 
+        capture_security_activity(
```
Comment (scefali) [inline src/sentry/api/endpoints/api_tokens.py:65]: @mdtro Might be useful code but why is this here?

File: src/sentry/api/invite_helper.py [hunk 2]
```diff
@@ -17,50 +16,51 @@
...
-        INVITE_COOKIE,
-        urlencode({"memberId": member_id, "token": token, "url": url}),
-        max_age=COOKIE_MAX_AGE,
-    )
+    Args:
+        request: rest_framework.request.Request
+        member_id: int
+        token: str
+    """
+    request.session["invite_token"] = token
```
Comment (scefali) [inline src/sentry/api/invite_helper.py:28]: @mdtro curious why we are doing it via session instead of a cookie. Is it security related?

**Comment by scefali** [comment]
@mdtro Would be good to have a description on why we are doing this. I thought the task was to make sure multiple people can't accept the same invitation? I'm guessing this is to prevent any problems where a cookie could be taken by a malicious actor but having this explained would be better.

File: src/sentry/api/invite_helper.py (hunk 2 above)
Comment (mdtro) [reply src/sentry/api/invite_helper.py:28]: Yes, this is security-related. The cookie value can be manipulated to allow for invite code reuse.

File: src/sentry/api/invite_helper.py [hunk 3]
```diff
@@ -17,50 +16,51 @@
...
 
-def remove_invite_cookie(request, response):
-    if INVITE_COOKIE in request.COOKIES:
-        response.delete_cookie(INVITE_COOKIE)
+def remove_invite_details_from_session(request):
+    """Deletes invite details from the request session"""
+    try:
+        del request.session["invite_member_id"]
+        del request.session["invite_token"]
+    except KeyError:
```
Comment (dashed) [inline src/sentry/api/invite_helper.py:37]: @mdtro A `KeyError` could occur for `invite_member_id`, but not for `invite_token`. And thus, `invite_token` would not be properly deleted.

I wonder if `request.session.pop("invite_member_id", None)` would be suitable.

File: src/sentry/api/invite_helper.py [hunk 4]
```diff
@@ -75,17 +75,17 @@ def from_cookie_or_email(cls, request, organization, email, instance=None, logge
         )
 
     @classmethod
-    def from_cookie(cls, request, instance=None, logger=None):
-        org_invite = get_invite_cookie(request)
+    def from_session(cls, request, instance=None, logger=None):
+        invite_token, invite_member_id = get_invite_details(request)
 
-        if not org_invite:
+        if not invite_token and not invite_member_id:
```
Comment (dashed) [inline src/sentry/api/invite_helper.py:81]:
```suggestion
        if not invite_token or not invite_member_id:
```

Using demorgan's law on `not (invite_token and invite_member_id)`

File: tests/sentry/api/endpoints/test_user_authenticator_enroll.py [hunk 5]
```diff
@@ -313,23 +311,24 @@ def assert_invite_accepted(self, response, member_id: int) -> None:
...
+        assert not self.client.session.get("invite_token")
+        assert not self.client.session.get("invite_member_id")
 
-    def setup_u2f(self):
+    def setup_u2f(self, om):
         new_options = settings.SENTRY_OPTIONS.copy()
         new_options["system.url-prefix"] = "https://testserver"
         with self.settings(SENTRY_OPTIONS=new_options):
+            # these session helper functions appear to modify the
+            # self.client.session?
```
Comment (dashed) [inline tests/sentry/api/endpoints/test_user_authenticator_enroll.py:322]: Should `self.client.session` not be modified here?

**Comment by dashed** [comment]
I'm 👍 on merging this into the request session.

File: src/sentry/api/invite_helper.py [hunk 6]
```diff
@@ -17,50 +16,37 @@
 from sentry.utils import metrics
 from sentry.utils.audit import create_audit_entry
 
-INVITE_COOKIE = "pending-invite"
-COOKIE_MAX_AGE = 60 * 60 * 24 * 7  # 7 days
```
Comment (dashed) [inline src/sentry/api/invite_helper.py:21]: @mdtro I was curious if we need to expire the invite details within 7 days of it being set on the request session. This would preserve a similar behaviour from when the invite details was stored as a cookie.

**Review by dashed — APPROVED** [review]
Looks good to me. I just had a question on how pending invites should be expired.

**Comment by corps** [comment]
Just to confirm -- it seems we still use signed cookies as the backend for our session store, so ultimately we're still using cookies, but they are more secured right?

**Comment by mdtro** [comment]
> Just to confirm -- it seems we still use signed cookies as the backend for our session store, so ultimately we're still using cookies, but they are more secured right?

Yes, that's correct. This puts them in the signed portion of the cookie, so if it's tampered with the SessionMiddleware will detect it and reject the request.

It also means that if we ever decide to switch to a different SessionEngine we won't have to modify this code.
